### PR TITLE
Fixes type promotion for `cat`

### PR DESCRIPTION
**Pull request by emcastillo**
Fixes a bug introduced in  #35030 

Changes the test to do all the possible type combinations.

### CODE REVIEW

**Review by albanD — APPROVED** [review]
LGTM

We want to fix this promptly so I'm going to merge this as-is.
If you want to do a follow up PR for the nits, add me as a reviewer, otherwise it's ok.

File: aten/src/ATen/native/TensorShape.cpp [hunk 1]
```diff
@@ -157,12 +160,13 @@ Tensor & _cat_out_cpu(Tensor& result, TensorList tensors, int64_t dim) {
     }
 
     if (tensor.sizes() != notSkippedTensor.sizes() ||
-        tensor.strides() != notSkippedTensor.strides() ||
-        tensor.dtype() != notSkippedTensor.dtype()) {
+        tensor.strides() != notSkippedTensor.strides()) {
```
[the remark this answers is not in the document]
Comment (emcastillo) [reply aten/src/ATen/native/TensorShape.cpp:163]: I split it in two, if you see there is a new if below that deals explicitly with type promotion using the `no_type_promotion` flag. I thought the behavior was clearer in this way.
I am fine with getting the old behavior back if you think its better.

File: aten/src/ATen/native/TensorShape.cpp [hunk 2]
```diff
@@ -139,6 +139,9 @@ Tensor & _cat_out_cpu(Tensor& result, TensorList tensors, int64_t dim) {
   // when the input tensors are of the same size and strides,
   // reuse the same iterator for all input tensors
   bool reuse_iterator = true;
+  bool no_type_promotion = true;
```
[the remark this answers is not in the document]
Comment (emcastillo) [reply aten/src/ATen/native/TensorShape.cpp:142]: sure! 😅

**Comment by emcastillo** [comment]
Thanks for the quick review and landing!
I did a minor follow-up :)

**Comment by albanD** [comment]
Sure. Ping me there when it's ready.

**Pull request closed without merging**